File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {
        int n = s.length();
        int res = 0, i = 0;
        boolean isPositive = true;
        
        while (i < n && s.charAt(i) == ' ') {
            i++;
        }
        
        if (i < n) {
            if (s.charAt(i) == '-') {
                isPositive = false;
                i++;
            } else if (s.charAt(i) == '+') {
                i++;
            }
        }
        
        while (i < n && s.charAt(i) >= '0' && s.charAt(i) <= '9') {
            if (res > Integer.MAX_VALUE / 10 || (res == Integer.MAX_VALUE / 10 && s.charAt(i) - '0' > 7)) {
                return isPositive ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            }
            res = res * 10 + (s.charAt(i) - '0');
            i++;
        }
        
        return res * (isPositive ? 1 : -1);
    }
}

File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.cpp
class Solution {
public:
    int f(int i, string& s, int res, int& N, int& sign) {
        if (i == N) {
            return sign * res;
        }
        
        if (s[i] < '0' || s[i] > '9') {
            return sign * res;
        }
        
        if ((res > INT_MAX / 10) || (res == INT_MAX / 10 && s[i] > '7')) {
            return sign == 1 ? INT_MAX : INT_MIN;
        }
        
        return f(i + 1, s, res * 10 + (s[i] - '0'), N, sign);
    }
    
    int myAtoi(string s) {
        int sign = 1;
        
        int i = 0, N = s.size();
        
        while (i < N && s[i] == ' ') {
            i++;
        }
        
        if (s[i] == '-') {
            sign = -1;
            i++;
        } else if (s[i] == '+') {
            i++;
        }
        
        return f(i, s, 0, N, sign);
    }
};